### ISSUE TITLE
poor error messages if a command line is rejected by .guard()

### ISSUE THREAD

**Issue by DanielT** [pacak/bpaf#113]
I'm trying to express some dependencies between the different options of my tool using .guard(), e.g. "you must use either --x or --y, but not both at the same time".

The first problem I have here is that this gives an error message that is prefixed with "Couldn't parse:". This is incorrect of course, the command line did get parsed into an options structure, my code merely rejected the result.

The second issue is that my whole reason for using .guard() was so that bpaf could then print a usage hint in addition to the error message, but it does not do this. If bpaf adds no information, then there is really no reason at all to use .guard().

For comparison, here is the output from bpaf 0.7.1:
```
Couldn't parse: You must either specify an INPUT file or use --create
```

This is the output from clap 2.33:
```
error: The following required arguments were not provided:
    <INPUT|--create>

USAGE:
    prog [FLAGS] [OPTIONS] <INPUT|--create> [--]

For more information try --help
```

**Comment by pacak** [comment]
Thank you for the report, I'll look into improving the error message. Do you have code for this publicly available by any chance?

**Comment by DanielT** [comment]
My code here (https://github.com/DanielT/a2ltool) currently uses clap 2.34.

My bpaf conversion isn't finished yet, but I just pushed my WIP branch. It is here: https://github.com/DanielT/a2ltool/tree/bpaf-commandline-wip

**Comment by pacak** [comment]
FWIW options like "you must use either --x or --y, but not both at the same time" are best expressed with `enum`, see some explanations here: https://pacak.github.io/bpaf/bpaf/_applicative/index.html#so-why-use-applicative-functors-then

> but I just pushed my WIP branch.

Good enough, will take a look. Might take a day or two. Error message with enum is probably not very good right now either but it should be possible to improve.

**Comment by pacak** [comment]
Published 0.7.2 which should help with the error message. As for the code - the main idea is to have mutually exclusive options represented as `enum`, as of 0.7.2 `bpaf` generates a better error message in that case.

So for using existing input or creating a new one you can do something like this:

```rust
enum Input {
    Create,
    Existing(PathBuf)
}

fn input() -> impl Parser<Input> {
    let create = long("create").req_flag(Input::Create);
    let existing = short('x').argument("FILE").map(Input::Existing);
    construct!([create, existing])
}
```

then use `input()` to parse your single input field.

Documentation should be clear enough about the representation, let me know if something is still confusing - I'll try to improve it.

**Comment by DanielT** [comment]
Thanks! The message is still terse, but much better.
As far as I'm concerned this fixes the issue.
